### ISSUE TITLE
NW & Electron Limitations

### ISSUE THREAD

**Issue by ronny1982** [manga-download/haruneko#621]
# NW

### User-Agent in Background Script not correctly applied
In newer versions of NW, the User-Agent in the background script (`navigator.userAgent`) defaults to `Node.js/21`.
This results in always triggering the restart notification when assigning a User-Agent, even when the real User-Agent didn't changed.
See also: https://github.com/nwjs/nw.js/issues/8162

Affected:
https://github.com/manga-download/haruneko/blob/1874bcf37dea7a20a5ceb9f00fcdeb932139e59c/app/nw/src/ipc/RemoteProcedureCallContract.ts#L27

### ...
...

# Electron

**Comment by MikeZeDev** [comment]
> New Browser Window is inaccessible

Is that a problem in NW.JS or in our code?

**Comment by ronny1982** [comment]
> > New Browser Window is inaccessible
> 
> Is that a problem in NW.JS or in our code?

It's very likely an issue in NW.js

**Comment by MikeZeDev** [comment]
Is there a possibility to hack your way to set user agent or referrer in NW js window?  

```ts
Object.defineProperty(document, "referrer", {get : function(){ return "my new referrer"; }});
Object.defineProperty(navigator, "userAgent", {get : function(){ return "My totally not suspicious browser"; }});
```

**Comment by ronny1982** [comment]
The question is to vague, a detailed explanation is required to answer that.

**Comment by MikeZeDev** [comment]
Case : TAADD, Tenmanga, NineManga (All of them but RU is the problem)

Manga : https://ru.ninemanga.com/manga/Контратака+изгнанного+ученика.html

Lets call NinemangaRU the SOURCE website.

1) On manga page, chapters links are normal : i.e  https://ru.ninemanga.com/chapter/Контратака%20изгнанного%20ученика/5310411.html

2) If you fetch this link WITHOUT  SOURCE  as Referrer, you'll stay on original website, and you have a chapter page, where you can work without problem.

![image](https://github.com/user-attachments/assets/c088e793-c223-404b-85d0-450b28e32fca)

You can see here its unusable on russian part of ninemanga as they miss all pics (404 errors)

2bis) If you fetch  this link WITH SOURCE  as Referrer, you get redirected to a dummy website (LINK A: https://www.financewealthloans.com/go/runm/5310411 ) where you have to chose a source server : 

![image](https://github.com/user-attachments/assets/c17e6d69-c73e-4b77-b49f-dea48de9ee83)

3) We will choose "Source 1", https://www.gardenhomefuture.com/go/jump/?type=runinemanga&cid=5310411. Lets call this url LINK B Again, we have to fetch this url using LINK A as Referrer.

The content is quite simple : 

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title></title>
    <META content="" name=Keywords>
    <META content="" name="Description">
    <link rel="shortcut icon" href="/files/images/yyzz/favicon.ico" type="image/x-icon">
    <meta property="og:type" content="website" /></head>
<body>
<script language="javascript" type="text/javascript">
    window.location.href="/statuses/2980106.html";
</script>
</body>
</html>
```

Nothing fancy. Its uses javascript to redirect to https://www.gardenhomefuture.com/statuses/2980106.html (Lets call it FINAL). 

4) FINAL link is displaying first picture of chapter. You can access all pictures url by reading a Javascript variable on the fly.

So, you will tell me : We just have to go through the process and inject a script on FINAL page, right? 
Heck, no.

Actually you need TWO THINGS to be able to properly get FINAL page : 
* use LINK B as Referrer
* a specific cookie telling that we are ok to display the chapter 

We cant set a custom referrer when we inject Javascript using NWJS. So we cant use JS injection.

Providing that the cookie is easy to forge using chapterid (**5310411**.html) and the id of FINAL page (**2980106**.html), we can
fetch FINAL url using it and proper REFERRER and.... parse the page HTML using regexp to get pages URL 🤦 

![image](https://github.com/user-attachments/assets/bbae20bd-930e-4e28-adc8-0cd830e287c7)

![image](https://github.com/user-attachments/assets/ff451125-4e1e-4d5c-a990-5da120483aa0)


Concrete problem : 

Sometimes we must execute Script in Window but in order to get correct content, the request used need a particular Referrer, and perhaps Cookies.

I had to parse the javascript using regexp to get what i wanted because no refererr change possible . I had to forge a cookie when it should have just being available. Ewww.

For now we cant set a Referrer on NWJS windows, i think we can with Electron windows. 

> win.loadURL(url[, options])
> url string
> options Object (optional)
> 
>     httpReferrer (string | [Referrer](https://www.electronjs.org/docs/latest/api/structures/referrer)) (optional) - An HTTP Referrer URL.
>     userAgent string (optional) - A user agent originating the request.
>     extraHeaders string (optional) - Extra headers separated by "\n"

I was wondering if it was possible to hack our way by monkey patching `document.referrer` getter? 

If we dig a big i think we can find similar cases like LectorManga (where you just CANT DO like we do with Tumanga, ive checked)

**Comment by MikeZeDev** [comment]
Another case  : https://github.com/manga-download/hakuneko/issues/7329

* Manga & chapter pages must be acceded with a mobile user agent. Otherwise >403 error. 
* Pages urls are available in a javascript variable.
* Since we cant set a referrer on NW window, we cant use FetchWindowScript to gather pages.

The alternative is to do manual fetch and parse par with regexp 🤮

**Comment by ronny1982** [comment]
No, the user-agent cannot be modified in new browser windows. This would also interfere with the global user-agent which is used to bypass CloudFlare. However, it is possible to inject fake with [FetchWindowPreloadScript](https://github.com/manga-download/haruneko/blob/25ad8969964ab28e06ba3607205e8691fbd12e3d/web/src/engine/platform/FetchProviderCommon.ts#L186)

**Comment by MikeZeDev** [comment]
Case  : [8comic](https://github.com/manga-download/hakuneko/issues/8299)

Problem :  No referer 

* chapter viewer url is hosted on another website and need 8comic referer to display pictures
* we could easily gather page urls using JS but opened windows dont have a referer (so no  FetchWindowScript)
* obfuscated JS on page is simple but decryption pattern is random / there are many... so using FetchRegex and manually decrypt stuff is a bad idea.

How would you do that? 
> 
>     Get the raw HTML content using Fetch with the corresponding referer
>     Open a new browser window for the corresponding URL
>     Render the previously acquired raw HTML in the browser window

**Comment by 27136773** [comment]
The referrer is assigned to the new Electron Browser Window based on the provided request.
Could this be a bug in electron?

https://github.com/manga-download/haruneko/blob/master/web/src/engine/platform/electron/RemoteBrowserWindow.ts#L75-L78

Try to open the following URL with FetchBrowser and verify the referer header
https://echo-http-requests.appspot.com/echo

## Edit
This is a bug when extracting the header to compose the `LoadURLOptions`. The current implementation uses `headers.get('Referer')`, but at this point the header is still concealed and therefore it should be `X-FetchAPI-Referer`.
A dirty workaround would be to use the `X-FetchAPI-Referer` header for the request.

## Edit
Addressed in 19616ecfa8e96cca6d461f0d78322162239805eb and 4a06a02481bff5585707b63ab46e97df7d0cd76a

**Comment by MikeZeDev** [comment]
Thanks, i will test that with Electron. sad problem may persist with nwjs tho.
Edit : works like a charm 👍
